### ISSUE TITLE
Please explain how the security works

### ISSUE THREAD

**Issue by fresheneesz** [bealearts/secure-storage-js#1]
If you want anyone to take a security library like this seriously, you can't just leave it to them to read and understand your code. Please explain all of the following:
- What algorithms this uses, and from where
- What algorithms this _implements_
- What possible attack vectors there are
- Any limitations
- Anything the developer using this library is required to do to ensure security
- Any recommendations related to the secure use of this library

You really need to explain why you think this is secure.

**Comment by Download** [comment]
The way I see it this is as secure as the secret key.
Now only I need some secure storage to store the key  :)

I would really like some explanation of how one could use this API to securely store an access token. Sure we could deliver the secure key to the client with an Ajax request.... but how to secure this request if  the token needed to secure it is the very thing we want to store?

**Comment by fresheneesz** [comment]
@Download The only secure way to do this would be to ask the user for a password.

**Comment by DavidBeale** [comment]
Yes, the idea would be for a password to be entered. @Download I would suggest looking at [JSON Web Tokens](https://jwt.io/) for you use case.

**Comment by Download** [comment]
I have come up with an idea for a solution. It requires a server, but this could be any server. I am prototyping it at the moment to play with it and see how it works. Would be interested in your thoughts on this.

Here's the idea:

You run a server configured with a secret key. This server answers POST and GET requests to some endpoint, say:

`https://example.com/secure-storage`

The client would gather some information needed to create a secure storage space and make a POST request to the server with that info:
* `name`: A name for the secure storage (unique for the domain) - can be in the JS source
* `client`: A client/device fingerprint. Look at [Client JS](https://clientjs.org/) for an example
* `expires-in`: The time in which the storage is valid, in seconds
* `nonce`: A random nonce string
* `origin`: The browser automatically adds an Origin header to the request

The server takes the incoming information and adds:
* `created-on`: The creation timestamp, in seconds since Unix epoch
* `cookie`: A server-side generated random string (which it sets as a HttpOnly cookie with the given expires time) 

The server then creates a hash based on all this info, using it's secret key. It returns the hash, setting the cookie in the response.

The client uses the hash as the password the user otherwise might have entered and stores the data encrypted in the store. The hash is never stored, but some of the fields are stored unencrypted:
* `nonce`
* `created-on`
* `expires-in`

On a next run, the client detect that a secure storage area already exists. It checks `created-on` and `expires-in` to determine if it's still valid. If not, it deletes the local storage area.

If the storage is still valid, the client makes a GET request to the server, adding this info:
* `name` - from JS source
* `client` - from device fingerprint
* `created-on` - from local storage
* `expires-in` - from local storage
* `nonce` - from local storage
* `origin`: The browser automatically adds an Origin header to the request
* `cookie`: The browser automatically adds the HttpOnly cookie

The server gets all the information it needs. And here is the key thing: the server validates the Origin domain and that the storage is not expired, before it generates and returns the hash. And people can't send wrong info here as it's part of the hash.

The client then uses the hash it got from the server as the password to decrypt the stored data.

This way, the info needed to get the encryption key is spread out over different places:

* The name is in the JS source, which means malware harvesting localStorage will not have it
* The client fingerprint is bound to the device
* The HttpOnly cookie cannot be read by JS, so cannot easily be stolen by malicious scripts
* The nonce is in localStorage itself
* The server secret key remains safe on the server

What do you guys think?
